### DISCUSSION TITLE
Is there a way to have debug code like print statements eliminated after debugging?

### DISCUSSION THREAD

**Discussion by yaconsult** [micropython/micropython#9683] [Using MicroPython]
What I mean is something like I have done before with the C preprocessor, for example, to easily include and exclude print debugging statements:

`#define DEBUG 0`
`#if DEBUG == 1`
`#define debug(x) Serial.print(x)`
`#define debugln(x) Serial.println(x)`
`#else`
`#define debug(x)`
`#define debugln(x)`
`#endif`

`int addOne(int inX){`
`    debug("Received value:");`
`    debugln(inX);`
`   ...`
`}`

Is there a way to do something similar in micropython?  If there was a debug constant in micropython and an `if debug` referencing that constant,  would micropython optimize the code out if debug was false or would it still be executed as normal?

Is a logging module or something else a better way to do something like this?

Or do people just manually strip out the debugging code they used when creating the code?  I'm just trying to figure out a good workflow for micropython.

Thanks!

**Comment by dlech** [comment]
There is a recent pull request that enables this scenario: https://github.com/micropython/micropython/pull/8970

And I suppose it could work in older versions with an const int instead of a bool (follow link for example).

**Reply by jimmo** [reply]
This feature is available in recent nightly/unstable builds, and @dlech is right -- in older versions, using a const int works too.

```
from micropython import const
_DEBUG_MODE = const(0)

if _DEBUG_MODE:
  print("this will be eliminated")
```

Note that constants are module-local, so there isn't a way to do this with a global flag. Accessing a variable across modules will never be optimised out.

**Comment by jimmo** [comment]
> If there was a debug constant in micropython

There _is_ a debug constant in MicroPython :)

```py
if __DEBUG__:
  print("this will be eliminated by the compiler at optimisation level 1 or higher.
```

You can set the optimisation level for the compiler with

```
>>> import micropython
>>> micropython.opt_level(N)
```

this will apply to any subsequent code compiled at runtime (i.e. if you put this at the top of main.py it will apply to any future imports).

Or if compiling with mpy-cross, with `-O[N]` as an mpy-cross argument.

**Comment by dhylands** [comment]
I normally do something like this:
```python
if DEBUG_MODE_ENABLED:
    def debug(..args...):
        ..whatever you want to do for debug output..
else:
    def debug(..args..):
        pass
```
Then you just have the overhead of a function call.

**Reply by privatefcagney** [reply]
> Then you just have the overhead of a function call.

And the overhead of preparing all the arguments for the function call, which might be none trivial.

`log(DEBUG, os.listdir())
`
will eat all the memory to create the directory list, and throw it away when the stub ignores all its arguments.
